pkg/beholder/auth.go

```diff
@@ -1,27 +1,145 @@
 package beholder
 
 import (
+	"context"
 	"crypto/ed25519"
+	"encoding/binary"
 	"fmt"
+	"time"
+
+	"google.golang.org/grpc/credentials"
+)
+
+var (
+	// authHeaderKey is the name of the header that the node authenticator will use to send the auth token
+	authHeaderKey = "X-Beholder-Node-Auth-Token"
+	// authHeaderVersion is the version of the auth header format
+	authHeaderVersion1 = "1"
+	authHeaderVersion2 = "2"
 )
 
-// authHeaderKey is the name of the header that the node authenticator will use to send the auth token
-var authHeaderKey = "X-Beholder-Node-Auth-Token"
+const DefaultAuthHeaderTTL = 1 * time.Minute
+
+type AuthHeaderProvider interface {
+	Credentials() credentials.PerRPCCredentials
+}
+
+// authHeaderPerRPCredentials is a PerRPCCredentials implementation that provides the auth headers
+type authHeaderPerRPCredentials struct {
+	privKey     ed25519.PrivateKey
+	lastUpdated time.Time
+	headerTTL   time.Duration
+	headers     map[string]string
+	version     string
+}
+
+// AuthHeaderProviderConfig configures AuthHeaderProvider
+type AuthHeaderProviderConfig struct {
+	HeaderTTL time.Duration
+	Version   string
+}
+
+func NewAuthHeaderProvider(privKey ed25519.PrivateKey, config *AuthHeaderProviderConfig) AuthHeaderProvider {
+	if config == nil {
+		config = &AuthHeaderProviderConfig{}
+	}
+	if config.HeaderTTL <= 0 {
+		config.HeaderTTL = DefaultAuthHeaderTTL
+	}
+	if config.Version == "" {
+		config.Version = authHeaderVersion2
+	}
+
+	creds := &authHeaderPerRPCredentials{
+		privKey:   privKey,
+		headerTTL: config.HeaderTTL,
+	}
+	// Initialize the headers ~ lastUpdated is 0 so the headers are generated on the first call
+	creds.refresh()
+	return creds
+}
+
+func (a *authHeaderPerRPCredentials) Credentials() credentials.PerRPCCredentials {
+	return a
+}
+
+func (a *authHeaderPerRPCredentials) GetRequestMetadata(_ context.Context, _ ...string) (map[string]string, error) {
+	return a.getHeaders(), nil
+}
+
+func (a *authHeaderPerRPCredentials) RequireTransportSecurity() bool {
+	return false
```

Comment:
should maybe make this configurable, even though our main use case is over localhost